WalletWasabi.Gui/Tabs/EncryptionManager/DecryptMessageViewModel.cs

```diff
@@ -88,22 +104,39 @@ public DecryptMessageViewModel(EncryptionManagerViewModel owner) : base("Decrypt
 				() =>
 				{
 					WarningMessage = "";
-					MyPublicKey = Global.WalletService.KeyManager.EncryptedSecret.GetSecret(Password).PubKey.ToHex();
-					IsPublicKeyPresent = true;
-				}
+					DecryptedMessage = DecryptMessage(EncryptedMessage, MyPublicKey, Password);
+				},
+				canDecrypt
 			);
 
 			MyPublicKeyCommand.ThrownExceptions.Subscribe(ex =>
 			{
 				WarningMessage = ex.Message;
 			});
+
+			this.WhenAnyValue(x => x.SelectedItem)
+				.Subscribe((x) => MyPublicKey = x is null ? "" : x.PubKey);
+
+			IEnumerable<HdPubKey> keys = Global.WalletService.KeyManager.GetKeys();
+
+			Addresses = new ObservableCollection<AddressPubKeyViewModel>(keys.Select(a => new AddressPubKeyViewModel(a)));
 		}
 
-		private static string DecryptMessage(string message, string password)
+		private string DecryptMessage(string message, string pubkey, string password)
 		{
 			password = Guard.Correct(password);
-			var bitcoinPrivateKey = Global.WalletService.KeyManager.EncryptedSecret.GetSecret(password);
-			return bitcoinPrivateKey.PrivateKey.Decrypt(message);
+			HdPubKey hdPubKey = Global.WalletService.KeyManager.GetKeys().FirstOrDefault(k => k.PubKey.ToHex() == pubkey);
```

Comment:
Instead of encoding each pubkey to hex to compare them with the one passed by parameter, I would recommend doing the opposite, create a Pubkey with the one hex passed by parameter and and compare pubkey instances (it is much faster).

Reply:
fixed!